core/finality-grandpa/src/communication/gossip.rs

```diff
@@ -430,58 +432,87 @@ type MaybeMessage<Block> = Option<(Vec<PeerId>, NeighborPacket<NumberFor<Block>>
 		Inner {
-			local_view: View::default(),
+			local_view: None,
 			peers: Peers::default(),
 			live_topics: KeepTopics::new(),
 			next_rebroadcast: Instant::now() + REBROADCAST_AFTER,
+			authorities: Vec::new(),
 			config,
 		}
 	}
 
-	/// Note a round in a set has started.
-	fn note_round(&mut self, round: Round, set_id: SetId) -> MaybeMessage<Block> {
-		if self.local_view.round == round && self.local_view.set_id == set_id {
-			return None;
-		}
+	/// Note a round in the current set has started.
+	fn note_round(&mut self, round: Round) -> MaybeMessage<Block> {
+		{
+			let local_view = match self.local_view {
+				None => return None,
+				Some(ref mut v) => if v.round == round {
+					return None
+				} else {
+					v
+				},
+			};
+
+			let set_id = local_view.set_id;
 
-		debug!(target: "afg", "Voter {} noting beginning of round {:?} to network.",
-			self.config.name(), (round, set_id));
+			debug!(target: "afg", "Voter {} noting beginning of round {:?} to network.",
+				self.config.name(), (round,set_id));
 
-		self.local_view.round = round;
-		self.local_view.set_id = set_id;
+			local_view.round = round;
 
-		self.live_topics.push(round, set_id);
+			self.live_topics.push(round, set_id);
+		}
 		self.multicast_neighbor_packet()
 	}
 
 	/// Note that a voter set with given ID has started. Does nothing if the last
 	/// call to the function was with the same `set_id`.
-	fn note_set(&mut self, set_id: SetId) -> MaybeMessage<Block> {
-		if self.local_view.set_id == set_id {
-			return None;
-		}
+	fn note_set(&mut self, set_id: SetId, authorities: Vec<AuthorityId>) -> MaybeMessage<Block> {
+		{
+			let local_view = match self.local_view {
+				ref mut x @ None => x.get_or_insert(View {
+					round: Round(0),
+					set_id,
+					last_commit: None,
+				}),
+				Some(ref mut v) => if v.set_id == set_id {
+					return None
+				} else {
+					v
+				},
+			};
 
-		self.local_view.update_set(set_id);
-		self.live_topics.push(Round(0), set_id);
+			local_view.update_set(set_id);
+			self.live_topics.push(Round(0), set_id);
+			self.authorities = authorities;
+		}
 		self.multicast_neighbor_packet()
 	}
 
 	/// Note that we've imported a commit finalizing a given block.
 	fn note_commit_finalized(&mut self, finalized: NumberFor<Block>) -> MaybeMessage<Block> {
-		if self.local_view.last_commit.as_ref() < Some(&finalized) {
-			self.local_view.last_commit = Some(finalized);
-			self.multicast_neighbor_packet()
-		} else {
-			None
+		{
+			match self.local_view {
+				None => return None,
+				Some(ref mut v) => if v.last_commit.as_ref() < Some(&finalized) {
+					v.last_commit = Some(finalized);
+				} else {
+					return None
+				},
+			};
 		}
+
+		self.multicast_neighbor_packet()
 	}
 
 	fn consider_vote(&self, round: Round, set_id: SetId) -> Consider {
-		self.local_view.consider_vote(round, set_id)
+		self.local_view.as_ref().map(|v| v.consider_vote(round, set_id))
+			.unwrap_or(Consider::RejectFuture)
```

Comment:
We should probably have a new status code `RejectUnknown` since when we don't have a local view we won't be broadcasting a neighbor packet, not sure we want to consider the cost to be the same as of a future message.